### PR TITLE
Fairness Metrics

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Changes proposed in this pull request:

- Created the fairness module and added four fairness metrics: `Independence`, `Separation`, `Sufficiency`, and `Demographic Parity Without Ground Truth`.

### CODE REVIEW

File: allennlp/fairness/fairness_metrics.py [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+
+It is provably impossible to satisfy any two of Independence, Separation, and Sufficiency simultaneously,
+except in degenerate cases.
+"""
+
+
+@Metric.register("independence")
+class Independence(Metric):
+    """
+    Independence. Assumes integer labels, with
```
Comment (AkshitaB) [inline allennlp/fairness/fairness_metrics.py:30]: It might be helpful to add a small example of how the output can be interpreted in a real-world context. Eg. kl div of nearly 0 for some protected variable implies independence, which is useful in a certain context.

File: allennlp/fairness/fairness_metrics.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        # There currently does not exist a robust loopless way to compute the conditional distributions
+        # Assumes num_protected_variables is small
+        for a in range(num_protected_variables):
+            C_given_a_dist = Categorical(
+                predicted_labels[protected_variable_labels == a]
+                .float()
+                .histc(bins=num_classes, min=0, max=num_classes - 1)
+                / predicted_labels.nelement()
+            )
+            kl_divs[a] = kl_divergence(C_given_a_dist, C_dist)
```
[the remark this answers is not in the document]
Comment (AkshitaB) [reply allennlp/fairness/fairness_metrics.py:90]: Also, this will compute the distributions correctly only for a single batch, right? We want that the distributions are computed based on all batches. (Again, you can take a look at other metrics for how this works). Ideally, the `__call__` function accumulates the distributions in some way, and the `get_metric()` function computes the final kl divergence. (I'm not sure how efficient this will be, though).

File: allennlp/fairness/fairness_metrics.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+from torch.distributions.categorical import Categorical
+from torch.distributions.kl import kl_divergence
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.training.metrics.metric import Metric
+
+"""
+Fairness metrics are based on:
+1) Barocas, S.; Hardt, M.; and Narayanan, A. 2019. Fairness and machine learning. fairmlbook.org.
+2) Zhang, B. H.; Lemoine, B.; and Mitchell, M. 2018. Mitigating unwanted biases with adversarial learning.
```
[the remark this answers is not in the document]
Comment (AkshitaB) [reply allennlp/fairness/fairness_metrics.py:13]: Also, we like semantic scholar links. 😄

**Review by AkshitaB — CHANGES_REQUESTED** [review]
@ArjunSubramonian I've left some comments. The main one is regarding the metrics working for different batches. When we run the trainer, we accumulate partial info for every batch, and compute the actual metric value for each epoch (and then reset). We want something similar here. Let me know if that makes sense, or if you would like to discuss it further.

File: allennlp/fairness/fairness_metrics.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+        predicted_labels: torch.Tensor,
+        protected_variable_labels: torch.Tensor,
+        num_classes: Optional[int] = None,
+    ) -> Dict[int, torch.FloatTensor]:
+        """
+        # Parameters
+
+        predicted_labels : `torch.Tensor`, required.
+            A tensor of predicted integer class labels of shape (batch_size, ...). Represented as C.
+        protected_variable_labels : `torch.Tensor`, required.
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/fairness_metrics.py:45]: Yes, that's correct

File: allennlp/fairness/fairness_metrics.py [hunk 5]
```diff
@@ -0,0 +1,921 @@
...
+        gold_labels : `torch.Tensor`, required.
+            A tensor of ground-truth integer class labels of shape (batch_size, ...). It must be the same
+            shape as the `predicted_labels` tensor. Represented as Y.
+        protected_variable_labels : `torch.Tensor`, required.
+            A tensor of integer protected variable labels of shape (batch_size, ...). It must be the same
+            shape as the `predicted_labels` tensor. Represented as A.
+        mask : `torch.BoolTensor`, optional (default = `None`).
+            A tensor of the same shape as `predicted_labels`.
+
+            Note: all tensors are expected to be on the same device.
```
Comment (epwalsh) [inline allennlp/fairness/fairness_metrics.py:245]: You could do format like this to make it look better on the docs site:

```suggestion

        !!! Note
            All tensors are expected to be on the same device.
```

File: allennlp/fairness/fairness_metrics.py [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
+4. Beutel, A.; Chen, J.; Zhao, Z.; and Chi, E. H. 2017. [Data decisions and theoretical implications when
+adversarially learning fair representations](https://api.semanticscholar.org/CorpusID:24990444).
+arXiv preprint arXiv:1707.00075.
+
+5. Aka, O.; Burke, K.; Bäuerle, A.; Greer, C.; and Mitchell, M. 2021.
+[Measuring model biases in the absence of ground truth](https://api.semanticscholar.org/CorpusID:232135043).
+arXiv preprint arXiv:2103.03417.
+
+It is provably impossible to satisfy any two of Independence, Separation, and Sufficiency simultaneously,
+except in degenerate cases.
```
Comment (epwalsh) [inline allennlp/fairness/fairness_metrics.py:23]: That's interesting. Do you have a reference for that?

File: allennlp/fairness/fairness_metrics.py (hunk 5 above)
Comment (ArjunSubramonian) [reply allennlp/fairness/fairness_metrics.py]: Woah, this is so cool :D I just changed it everywhere

File: allennlp/fairness/fairness_metrics.py (hunk 6 above)
Comment (ArjunSubramonian) [reply allennlp/fairness/fairness_metrics.py]: Added a link in the doctoring!

File: allennlp/fairness/fairness_metrics.py [hunk 7]
```diff
@@ -0,0 +1,940 @@
...
+            return {
+                x: torch.full((self._num_classes,), float("nan"))
+                for x in range(self._num_protected_variable_labels)
+            }
+
+        device = self._y_counts.device
+        pmi_terms = {}
+        prob_y = torch.zeros(self._num_classes).to(device)
+        torch.div(self._y_counts, self._total_predictions, out=prob_y)
+        for x in range(self._num_protected_variable_labels):
```
Comment (AkshitaB) [inline allennlp/fairness/fairness_metrics.py:917]: I think it's possible to vectorize this part if `joint_counts_by_protected_variable_label` and `_protected_variable_label_counts` are initialized as tensors instead of dicts (they can still be filled using the for loop like now).

Comment (ArjunSubramonian) [reply allennlp/fairness/fairness_metrics.py]: Great suggestion :) I just did this for all the metrics.

**Review by AkshitaB — APPROVED** [review]
